language/move-vm/runtime/src/interpreter.rs

```diff
@@ -869,6 +1033,504 @@ impl Frame {
+            }
+            Bytecode::ReadRef => {
+                let ref_ty = interpreter.operand_stack.pop_ty()?;
+                match ref_ty {
+                    Type::Reference(inner) | Type::MutableReference(inner) => {
+                        check_ability(resolver.loader().abilities(&inner)?.has_copy())?;
+                        interpreter.operand_stack.push_ty(inner.as_ref().clone())?;
+                    }
+                    _ => {
+                        return Err(PartialVMError::new(
+                            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                        )
+                        .with_message("ReadRef expecting a value of reference type".to_string()))
+                    }
+                }
+            }
+            Bytecode::WriteRef => {
+                let ref_ty = interpreter.operand_stack.pop_ty()?;
+                let val_ty = interpreter.operand_stack.pop_ty()?;
+                match ref_ty {
+                    Type::MutableReference(inner) if *inner == val_ty => {
+                        check_ability(resolver.loader().abilities(&inner)?.has_drop())?;
+                    }
+                    _ => {
+                        return Err(PartialVMError::new(
+                            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                        )
+                        .with_message("WriteRef expecting a value of mutable type".to_string()))
+                    }
+                }
+            }
+            Bytecode::CastU8 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U8)?;
+            }
+            Bytecode::CastU16 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U16)?;
+            }
+            Bytecode::CastU32 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U32)?;
+            }
+            Bytecode::CastU64 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U64)?;
+            }
+            Bytecode::CastU128 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U128)?;
+            }
+            Bytecode::CastU256 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U256)?;
+            }
+            Bytecode::Add
+            | Bytecode::Sub
+            | Bytecode::Mul
+            | Bytecode::Mod
+            | Bytecode::Div
+            | Bytecode::BitOr
+            | Bytecode::BitAnd
+            | Bytecode::Xor
+            | Bytecode::Or
+            | Bytecode::And => {
+                let lhs = interpreter.operand_stack.pop_ty()?;
+                let rhs = interpreter.operand_stack.pop_ty()?;
+                lhs.check_eq(&rhs)?;
```

Comment:
So we don't check the types for primitive value operations, including integer operations, boolean operations and comparision operations. The reason why they were safe is because our value system have separate representation between primitive values and struct values, so had there been a type confusion they would be caught at the runtime. I can either add a test to make sure it doesn't regress or I can add the checks here. What would be your preference?

Reply:
Either sounds great to me--as long as something will break if we switch to a different value representation in the future. Thanks!